src/recordlinker/schemas/algorithm.py

```diff
@@ -44,6 +44,86 @@ def serialize_func(self, value: matchers.FeatureFunc) -> str:
         return str(value)
 
 
+class LogOdd(pydantic.BaseModel):
+    """
+    The schema for an LogOdd record.
+    """
+
+    model_config = pydantic.ConfigDict(from_attributes=True)
+
+    feature: Feature = pydantic.Field(json_schema_extra={"enum": Feature.all_options()})
+    value: Annotated[float, pydantic.Field(ge=0)]
+
+    @pydantic.field_validator("feature", mode="before")
+    def validate_feature(cls, value):
+        """
+        Validate the feature is a valid PII feature.
+        """
+        try:
+            return Feature.parse(value)
+        except ValueError as e:
+            raise ValueError(f"Invalid feature: '{value}'. {e}")
+
+
+class SkipValue(pydantic.BaseModel):
+    feature: str = pydantic.Field(json_schema_extra={"enum": Feature.all_options() + ["*"]})
+    values: list[str] = pydantic.Field(min_length=1)
+
+    @pydantic.field_validator("feature", mode="before")
+    def validate_feature(cls, value):
+        """
+        Validate the feature is a valid PII feature.
+        """
+        if value == "*":
+            return value
+        try:
+            Feature.parse(value)
+        except ValueError as e:
+            raise ValueError(f"Invalid feature: '{value}'. {e}")
+        return value
+
+
+class AlgorithmContext(pydantic.BaseModel):
+    """
+    The schema for an algorithm context record.
+    """
+
+    model_config = pydantic.ConfigDict(from_attributes=True)
+
+    include_multiple_matches: bool = True
+    log_odds: typing.Sequence[LogOdd] = []
+    skip_values: typing.Sequence[SkipValue] = []
+
+    @pydantic.model_validator(mode="after")
+    def init_log_odds_helpers(self) -> typing.Self:
+        """
+        Initialize cache helpers for returning log odds values.
+        """
+        self._log_odds_cache: dict[str, float | None] = {}
+        self._log_odds_mapping: dict[str, float] = {str(o.feature): o.value for o in self.log_odds}
+        return self
+
+    def get_log_odds(self, value: Feature | BlockingKey) -> float | None:
+        """
+        Get the log odds for a specific Feature or BlockingKey.
+        """
+        key = str(value)
+        result: float | None = None
+
+        result = self._log_odds_cache.get(key, None)
+        if result:
+            return result
+
+        vals = value.values_to_match() if isinstance(value, Feature) else [str(value)]
+        for val in vals:
+            result = self._log_odds_mapping.get(val, None)
+            if result:
+                break
+
+        self._log_odds_cache[key] = result
```

Comment:
@bamader @m-goggins I could go either way on this cache key, but am starting to lean towards removing it.  The block above (117-121) is mostly just an operation of converting an object to a string and looking it up in a map (there is an edge case for identifiers).  Thoughts?

Reply:
If I'm understanding correctly, the edge case would be having multiple identifiers whose values you can retrieve from the cache instead of recomputing?

Reply:
I think I'm following. This line basically just updates the K-V pair in the caching dictionary so that if we did have to do an identifier lookup in the edge case you outlined, on future calls, we don't have to walk through that process again if the user gives us IDENTIFIER:DL but we only have IDENTIFIER. Is that right?

If that's the case, then it seems like keeping it isn't harmful, just a small quality of life thing that means we won't have to re-search the values in future calls (which I think would be likely to happen if the user doesn't update their log-odds or their algorithm, and neither of those should be changing very frequently). 

I definitely don't feel strongly, and if my thinking is correct, I may lean slightly towards keeping it but would be totally fine either way!

Reply:
Yeah it doesn't add much.  Most of the time `vals` declared on line 117 is going to be a list of 1 item.  With IDENTIFIER:XX it will be 2 items.  So in that case, the cache helps cut down on the lookups.  It's negligible though, so I'm inclined to remove it just so it reads better.

Reply:
That works for me. Readability is always one of my top factors, so I'm good to remove 👍